### PR TITLE
win_file: Added link support and general cleanup

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
To bring the windows module up to feature parity with the unix ones this PR is designed to add link handling to win_file. It also cleans up a few sections around it not setting changed to True when a change actually happened and lots more tests to test check mode and more scenarios.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
win_file

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 8f4652ab2f) last updated 2017/03/30 16:18:30 (GMT +1000)
  config file = /mnt/c/dev/test/role/ansible.cfg
  configured module search path = [u'/mnc/c/dev/test/role/library']
  python version = 2.7.12 (default, Dec 14 2016, 21:05:07) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION
This PR is meant to bring link functionality to win_file as it exists in the file module. There are some differences between the 2 due to differences in the platform. I'm not the most adept at C# code but it seems to do the trick and outputs what is required but appreciate any tips or pointers that you may notice.

### CODE REVIEW

File: lib/ansible/modules/windows/win_file.py
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_file.py]: With this update, it would manipulate links, so this line should change.

**Comment by jhawkesworth** [comment]
Would be great to have this.  Feel free to add yourself as a module author and/or maintainer.

I am out of touch on what the latest state is regarding shared code, but I wonder if there is any mileage in having win_stat and win_file share the link/junction code?

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: I'll change this tomorrow when I've got access to a computer, thanks for the pickup.

**Comment by jborean93** [comment]
I was talking to @nitzmahone and mentioned this possibility at the last WUG. I don't want to add this to the powershell.ps1 file due to the sheer amount of code that need to be run for every module. I said it would be great to have modular files we can reference as my ultimate goal was to use this code for win_stat and win_find. I know I plan on using this to fix https://github.com/ansible/ansible/issues/21048

**Comment by jhawkesworth** [comment]
Agreed powershell.ps1 is not the place for it.  I'm hoping we can start using more than just powershell.ps1 for shared stuff although I'm not sure if it gets sent across the wire every time a module is executed with the new execution logic now, so might be ok to put this (and the SID lookup stuff which there are a bunch of PRs floating around for) into _somewhere common_.  I'm sure @nitzmahone will let us know when this is viable and we can figure out what would make the most sense in which file.

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Have been updated

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.py]: Just a matter of opinion, I don't like the strings symbolic_link, junction_point and hard_link.
I would prefer: *symlink*, *hardlink* and *something* else for junction_point :-)

The strings for Linux/Unix is *link* and *hard* which is also not optimal ;-)

Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Happy to change I would probably lean more towards *link*, *hard*, *junction* for symbolic links, hard links and junctions points to keep things consistent with file. Happy to go with whatever but will add it to the next user group agenda.

**Comment by jborean93** [comment]
@dagwieers @nitzmahone I've changed the state to be what was agreed in the WUG and fixed up the tests. Over to you guys for any more reviews.

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.py]: I don't think the new description is any more clear. The module does manage files, not just attributes.
(So the original file module description is suboptimal IMO)

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.py]: Listing the aliases in the description is unneeded, we have `aliases:` for that.

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.py]: Nowadays, we do:
```yaml
  type: bool
  default: 'no'
```

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: No worries will update the description to be a bit better

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Is there a link or document that states this, it's a bit frustrating trying to keep up with these standards sometimes.

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Another copy and paste strikes again, I'll fix it up.

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [reply lib/ansible/modules/windows/win_file.py]: I feel your pain. @gundalow recently told me and it was in the documentation: https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#documentation-block

Comment (dagwieers) [reply lib/ansible/modules/windows/win_file.py]: Best thing to me is to update all the modules to comply with this, so every module is implicitly an example of how to do it.

Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Thanks for that, looks like you have set `version_added` in the parameters which I didn't know before hand so I'll update that at the same time.

File: lib/ansible/modules/windows/win_file.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.py]: This should be 'no', because otherwise it will be a boolean in the documentation and listed as False.
That is why `type: bool` translates to `choices: [ 'no', 'yes' ]`.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_file.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: So add in the quotes to the no?

Comment (dagwieers) [reply lib/ansible/modules/windows/win_file.py]: Yup, so people really should be adding no or yes without quotes. We want the "no" and "yes" be listed in the documentation as-is, and not as False or True. (Which no would translate to in YAML, unless it is quoted).

Comment (jborean93) [reply lib/ansible/modules/windows/win_file.py]: Cool, committed the changes.

**Review by dagwieers — APPROVED** [review]
shipit

**Comment by dagwieers** [comment]
shipit

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: Minor nit- since we've got .NET 4, can use `attr.HasFlag(FileAttributes.ReparsePoint)` for these...

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: HasFlag

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: Pretty sure `Process.GetCurrentProcess().SafeHandle.DangerousGetHandle()` gets you the same handle sans the P/Invoke def.

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: Maybe use builtin SafeHandle from .NET Process obj instead of custom PInvoke

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: HasAttr

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: `switch(type)` would be the idiomatic C# way...

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: HasAttr

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: A tighter pattern for all the boolean Win32 calls (since this is a *lot* of code we're shipping to the target each time- balance between readability/brevity is key):
```
if(!SomeWin32BooleanCall())
    throw new Win32Exception(Marshal.GetLastWin32Error());
```

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: Isn't this supposed to be `\\?\`?

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: s/32767/255

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1:450]: Docs say to pass NULL for lpOutBuffer on this: https://msdn.microsoft.com/en-us/library/windows/desktop/aa364595(v=vs.85).aspx

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1:446]: At first glance, all the usages of `REPARSE_DATA_BUFFER` appear to be blittable- since we're not using DeviceIoControl for anything else, we could just change the arg to `ref REPARSE_DATA_BUFFER` and ditch all the manual allocation/free stuff entirely with some tweaks to the struct.

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_file.ps1]: `$file.Attributes.HasFlag([System.IO.FileAttributes]::ReparsePoint)` should also work here.

File: test/integration/targets/win_file/tasks/main.yml
Comment (nitzmahone) [inline test/integration/targets/win_file/tasks/main.yml]: Should be able to single-quote all these and avoid the double-backslash escaping

**Review by nitzmahone — CHANGES_REQUESTED** [review]
There are some minor issues with the PInvoke type defs (and ways we could probably get rid of a lot of code there). A few other minor things noted as well. None of these necessarily need to hold up the merge, especially if you're not sure what I mean about the PInvoke stuff, but it'd be good to tighten some of this up if possible. We could also go over it in more detail together at some point if you want...

**Comment by jborean93** [comment]
Thanks @nitzmahone I'll have a shot at going through your comments, been a bit held up unfortunately.

File: lib/ansible/modules/windows/win_file.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1]: This is a poor choice for variable naming, `\??\` is used as a prefix when creating a ReparsePoint. There isn't really much documentation I could find from Microsoft but here is some info about it https://stackoverflow.com/questions/14329653/does-the-substitutename-string-in-the-pathbuffer-of-a-reparse-data-buffer-stru.

I'll move this down to the actual method used as it is only used once and `MAXIMUM_PATH_FIX_PREFIX` isn't really what it is used for.

File: lib/ansible/modules/windows/win_file.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1]: Are you saying to remove the custom error messages and just throw the Win32Exception wherever I've done this? I would have thought the trade off of having better error messages outweighs readability of a line.

File: lib/ansible/modules/windows/win_file.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1:450]: The `DeviceIoControl` currently being used here is expecting  `REPARSE_DATA_BUFFER` so I couldn't use `IntPtr.Zero` and just used an empty struct. Should I have 2 P/Invoke methods for DeviceIoControl for the different control codes or just keep as is?

File: lib/ansible/modules/windows/win_file.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1:446]: Sorry you've lost me there, it might be best if you make this change in another PR or be prepared for me to ask you numerous questions and try and find examples.

**Comment by jborean93** [comment]
Hey @nitzmahone I've made the changes requested except for the 3 I replied to. Let me know your thoughts on them.

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [reply lib/ansible/modules/windows/win_file.ps1]: No- the places where you've got custom messaging are good- just getting rid of the `success = SomeMethodCall(); if(!success)` pattern (collapsing that into `if(!SomeMethodCall()) throw`) since the utility of inspecting the `success` var in a debugger is pretty limited. Different story if it were an error code or something, but just as a bool, you know if you drop into the block, it's because it was false.

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [reply lib/ansible/modules/windows/win_file.ps1:450]: Ah, old P/Invoke tricks: because it's blittable, you should be able to change the definition to `class REPARSE_DATA_BUFFER` instead of `struct`- then you can pass null. :)

File: lib/ansible/modules/windows/win_file.ps1
Comment (nitzmahone) [reply lib/ansible/modules/windows/win_file.ps1:446]: I need to get the 2.3.2 stuff knocked out, then I can show you what I mean (probably with a maintainer commit- do you have that turned on for this PR?)

Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1:446]: yep, I'll make the changes you've requested and rebase with the latest commits as well.

File: lib/ansible/modules/windows/win_file.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_file.ps1:450]: I tried this out but I couldn't get it working, I'll wait until you've got some time after you 2.3.2 stuff like the comment above.

**Comment by jhawkesworth** [comment]
Just been testing  this on my laptop against s2008r2, win10 (host) s2012r2, s2016. 

Everything ran through without a hitch but it took >16 mins (albeit for 166 tasks). 
In comparison current devel has only 50 tasks took less than 5 mins.  Presumably CI hardware is quicker so perhaps not a problem but may see an increase in testing times once this is merged.

I can't really comment meaningfully on the C# code but it certainly seems to work.

shipit

**Comment by nitzmahone** [comment]
Yeah, I definitely don't want to turn away test coverage because it'll increase the CI times (within reason, of course)- we're playing around with different ways to segment that to minimize the number of modules that need to be tested on platform changes (eg module_utils). I also did some messing around with PS module integration tests using Pester... Especially for stuff like this, I could see it as a viable alternative, since in most cases the actual execution time of this module is dwarfed by the transfer overhead (so doing it in one shot via Pester could take a few seconds instead of minutes). Matt Clay and I will be doing some work around the end of 2.4 to integrate Pester code coverage, and I might try to slide Pester integration testing onto the pile as well.

**Comment by jhawkesworth** [comment]
Definitely wouldn't want to miss on the tests either, but smarter/faster ways to test the code would be great.  Hope the pester stuff works out. Once we have shared code I think we will feel the need to be able to do more than just integration test more keenly.

**Comment by jborean93** [comment]
Will be updating the branch and creating a new PR once ready.

**Pull request closed without merging**
